capactio/capact
# Implement concurrent manifest validation against Hub in CLI

## Description

Changes proposed in this pull request:

- Add JSON schema validation for Type
- Validate manifest references for Interface, Implementation and Type
- Refactor validator to concurrent validator
- Add verbose flag for validation
- Move `/pkg/sdk/manifest` validation-related logic to `/pkg/sdk/manifest/validation`

## Notes

This PR is based on branch with fixes from #436. Once #436 is merged, this PR will be rebased to latest `main`.
**EDIT:** Done

## Testing

1. Run Capact from `main` branch:
    ```bash
    make dev-cluster
    ```
1. Populate Hub manifests from `main` branch:
    ```bash
    kubectl -n capact-system port-forward svc/neo4j-neo4j 7687:7687
    APP_JSONPUBLISHADDR=http://$(ipconfig getifaddr en0) go run cmd/populator/main.go register ocf-manifests ../hub-manifests
    ```
1. Log in to Capact CLI: https://capact.io/docs/cli/getting-started#first-use
1. Run the following command against cloned `hub-manifests` in `main` branch:

    ```bash
    go run cmd/cli/main.go manifest validate --server-side -v ../hub-manifests/**/*.yaml
    ```

   - check the output with and without `-v / --verbose` flag
   - check the output with and without `--server-side` flag
   - press CTRL+C in the middle of validation to see how it behaves

1. Check out the fixed manifests from PR: https://github.com/capactio/hub-manifests/pull/16
1. Run the populator:
    ```bash
    kubectl -n capact-system port-forward svc/neo4j-neo4j 7687:7687
    APP_JSONPUBLISHADDR=http://$(ipconfig getifaddr en0) go run cmd/populator/main.go register ocf-manifests ../hub-manifests
    ```
1. Follow the 3rd step again. Observe no errors in manifests.

## Related issue(s)

#398
